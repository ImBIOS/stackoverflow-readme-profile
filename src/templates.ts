import * as path from 'path'
import Handlebars from 'handlebars'

import { Badges } from './fetch'
import { getTemplate, getTruncatedText, replaceAll } from './utils'
import { Theme as ThemeObject } from './index'

import defaultTheme from './themes/default.json'
import darkTheme from './themes/dark.json'
import cobaltTheme from './themes/cobalt.json'
import monokaiTheme from './themes/monokai.json'
import graywhiteTheme from './themes/graywhite.json'
import hotdogTheme from './themes/hotdog.json'

export type Theme = 'default' | 'dark' | 'cobalt' | 'monokai' | 'graywhite' | 'hotdog'
export interface ErrorTemplateContext {
  lines: string[]
}
export interface ProfileTemplateContext {
  avatar: string
  username: string
  reputation: string
  badges: Badges
  location?: string
  website?: string
  theme: ThemeObject
  badgesMarginLeft: number
  badgeSilverMarginLeft: number
  badgeBronzeMarginLeft: number
}
export interface ProfileParams extends Pick<ProfileTemplateContext,
  | 'avatar'
  | 'username'
  | 'reputation'
  | 'badges'
  | 'location'
  | 'website'
> {
  theme: Theme
}

export const THEMES: {[key in Theme]: ThemeObject} = {
  default: defaultTheme,
  dark: darkTheme,
  cobalt: cobaltTheme,
  monokai: monokaiTheme,
  graywhite: graywhiteTheme,
  hotdog: hotdogTheme
}
const LETTER_WIDTH = 9
const LETTER_MARGIN = 12

const PATH_TO_ERROR_TEMPLATE = path.resolve(__dirname, './templates/error.hbs')
const PATH_TO_TEMPLATE = path.resolve(__dirname, './templates/profile.hbs')
const PATH_TO_SO_ICON = path.resolve(__dirname, './templates/so-icon.hbs')
const PATH_TO_REP_BADGES_TEMPLATE = path.resolve(__dirname, './templates/reputation-badges.hbs')

const errorTemplate = getTemplate<ErrorTemplateContext>(PATH_TO_ERROR_TEMPLATE)

Handlebars.registerHelper('truncatedText', getTruncatedText)

const soIconTemplate = getTemplate(PATH_TO_SO_ICON)
Handlebars.registerPartial('so-icon', soIconTemplate)

const repBadgesTemplate = getTemplate(PATH_TO_REP_BADGES_TEMPLATE)
Handlebars.registerPartial('rep-badges', repBadgesTemplate)

const profileTemplate = getTemplate<ProfileTemplateContext>(PATH_TO_TEMPLATE)

export const renderProfile = (params: ProfileParams): string => {
  const { reputation, badges } = params
  const { gold, silver } = badges

  const badgesMarginLeft = LETTER_WIDTH * reputation.length + LETTER_MARGIN
  let badgeSilverMarginLeft = (LETTER_WIDTH * gold.toString().length)
  if (gold > 0) badgeSilverMarginLeft += LETTER_MARGIN

  let badgeBronzeMarginLeft = (LETTER_WIDTH * silver.toString().length)
  if (silver > 0) badgeBronzeMarginLeft += badgeSilverMarginLeft + LETTER_MARGIN

  return profileTemplate({
    ...params,
    theme: {
      ...defaultTheme,
      ...THEMES[params.theme]
    },
    badgesMarginLeft,
    badgeSilverMarginLeft,
    badgeBronzeMarginLeft
  })
}

export const renderError = (params: {error: string}): string => {
  const { error } = params

  // remove nested error title
  const errorMsg = replaceAll(error, 'Error: ', '')
  const lines = `Error: ${errorMsg}`.match(/.{1,45}/g) || [error]
  return errorTemplate({ lines })
}
