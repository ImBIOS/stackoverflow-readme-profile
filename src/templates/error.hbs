<svg version="1.1" baseProfile="full" width="300" height="86" viewBox="0 0 300 86" preserveAspectRatio="xMinYMin meet" xmlns="http://www.w3.org/2000/svg" style="overflow: visible;">
  <style>
    svg {
      overflow: visible;
    }
    .ctn {
      x: 0;
      y: 0;
      fill: #F8D7DB;
      stroke-width: 2;
      stroke: #AD6E75;
    }
    text {
      fill: #AD6E75;
      font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif;
    }
  </style>

  <rect
    width="300"
    height="86"
    class="ctn"
    rx="5"
    ry="5"
  ></rect>
  <text x="11" y="11">
    {{#each lines}}
      <tspan x="11" dx="6" dy="12">{{this}}</tspan>
    {{/each}}
  </text>
</svg>
