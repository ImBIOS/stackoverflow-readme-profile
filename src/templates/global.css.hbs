:root {
  --color-text: {{theme.colorText}};
  --color-bg: {{theme.colorBg}};
  --color-link: {{theme.colorLink}};
  --color-primary: {{theme.colorPrimary}};
  --color-gold: {{theme.colorGold}};
  --color-silver: {{theme.colorSilver}};
  --color-bronze: {{theme.colorBronze}};
  --color-so-icon: {{theme.colorSoIcon}};
}
@keyframes fadeIn {
  0% {opacity:0;}
  100% {opacity:1;}
}
svg {
  overflow: visible;
}
.animate {
  animation: fadeIn .5s forwards;
  opacity: 0;
}
text {
  fill: {{theme.colorText}};
  font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif;
}
.ctn {
  x: 0;
  y: 0;
  fill: var(--color-bg);
  stroke-width: 2;
  stroke: var(--color-primary);
}