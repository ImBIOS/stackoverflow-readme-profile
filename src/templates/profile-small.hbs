<svg version="1.1" baseProfile="full" width="240" height="57" viewBox="0 0 240 55" preserveAspectRatio="xMinYMin meet" xmlns="http://www.w3.org/2000/svg" style="overflow: visible;">
  <style>
    {{> global-styles }}
    .username {
      font-weight: 500;
      font-size: 16px;
    }
  </style>

  <rect
    width="240"
    height="57"
    class="ctn"
    rx="5"
    ry="5"
  ></rect>

  {{> so-icon translateX="195px" width="40px" }}

  <svg x="5" y="5">

    {{! -- START user avatar -- }}
     <rect id="imageCtn" transform="translate(0,0)" width="47" height="47"  rx="5"/>
      <clipPath id="clip">
        <use href="#imageCtn"/>
      </clipPath>
    </defs>
    <use href="#imageCtn"/>
    <image width="47" height="47" clip-path="url(#clip)" href="{{avatar}}" />
    {{! -- EMD user avatar -- }}

    <g transform="translate(55,20)">
      <text class="username animate" transform="translate(0,0)">{{username}}</text>
      <g class="animate" transform="translate(0,20)" style="font-weight: 800; animation-delay: 0.1s;">
        {{> rep-badges reputation=reputation badges=badges }}
      </g>
    </g>
  </svg>
</svg>
