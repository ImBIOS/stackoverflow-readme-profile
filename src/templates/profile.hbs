<svg version="1.1" baseProfile="full" width="300" height="86" viewBox="0 0 300 86" preserveAspectRatio="xMinYMin meet" xmlns="http://www.w3.org/2000/svg" style="overflow: visible;">
  <style>
    :root {
      --color-text: {{theme.colorText}};
      --color-bg: {{theme.colorBg}};
      --color-link: {{theme.colorLink}};
      --color-primary: {{theme.colorPrimary}};
      --color-gold: {{theme.colorGold}};
      --color-silver: {{theme.colorSilver}};
      --color-bronze: {{theme.colorBronze}};
    }
    @keyframes fadeIn {
      0% {opacity:0;}
      100% {opacity:1;}
    }
    svg {
      overflow: visible;
    }
    .animate {
      animation: fadeIn .5s forwards;
      opacity: 0;
    }
    .ctn {
      x: 0;
      y: 0;
      fill: var(--color-bg);
      stroke-width: 2;
      stroke: var(--color-primary);
    }
    text {
      fill: {{theme.colorText}};
      font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif;
    }
    .username {
      font-weight: 500;
      font-size: 16px;
    }
    .website, .location {
      font-size: 10px;
    }
    .website {
      fill: var(--color-link);
    }
  </style>

  <rect
    width="300"
    height="86"
    class="ctn"
    rx="5"
    ry="5"
  ></rect>
  <svg x="11" y="11" width="auto">

    {{! -- START user avatar -- }}
     <rect id="imageCtn" transform="translate(0,0)" width="64" height="64"  rx="5"/>
      <clipPath id="clip">
        <use href="#imageCtn"/>
      </clipPath>
    </defs>
    <use href="#imageCtn"/>
    <image width="64" height="64" clip-path="url(#clip)" href="{{avatar}}" />
    {{! -- EMD user avatar -- }}

    <g transform="translate(75,13)">
      <text class="username animate" transform="translate(0,0)">{{username}}</text>
      <g class="animate" transform="translate(0,22)" style="font-weight: 800; animation-delay: 0.1s;">
        {{> rep-badges reputation=reputation badges=badges }}
      </g>
      {{#if location}}
        <g class="animate" style="animation-delay: 0.2s;" transform="translate(0,38)">
          <text class="location">{{location}}</text>
        </g>
      {{/if}}
      {{#if website}}
        <g class="animate" style="animation-delay: 0.3s;" transform="translate(0,50)">
          <text class="website">{{website}}</text>
        </g>
      {{/if}}
    </g>
  </svg>

  {{> so-icon }}
</svg>
