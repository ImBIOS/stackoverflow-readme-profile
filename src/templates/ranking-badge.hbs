<svg version="1.1" baseProfile="full" width="{{ctnWidth}}" height="34" viewBox="0 0 {{ctnWidth}} 34" preserveAspectRatio="xMinYMin meet" xmlns="http://www.w3.org/2000/svg" style="overflow: visible;">
  <style>
    {{> global-styles }}

    :root {
      --color-so-icon: var(--color-primary);
    }

    text {
      font-weight: 500;
      font-size: 17px;
    }
    .ctn {
      stroke-width: 4;
    }
    .tag {
      fill: var(--color-link);
      font-weight: 700;
    }
  </style>

  <rect
    width="{{ctnWidth}}"
    height="34"
    class="ctn"
    rx="5"
    ry="5"
  ></rect>

  {{> so-icon translateX="5px" width="20px" }}

  <text class="animate" transform="translate(32,22)">
    Top {{rank}}%
    <tspan class="tag">{{tagName}}</tspan>
    users
  </text>
</svg>
