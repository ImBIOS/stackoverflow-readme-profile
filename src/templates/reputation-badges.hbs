<style>
  .reputation {
    font-weight: 600;
    font-size: 14px;
  }
  .badge-count circle {
    cx: 0;
    cy: -4;
    r: 4;
  }
</style>

<text class="reputation">{{reputation}}</text>
<g transform="translate({{badgesMarginLeft}}, 0)">
  {{#if badges.gold}}
    <g class="badge-count">
      <circle fill="var(--color-gold)"></circle>
      <text x="7">{{badges.gold}}</text>
    </g>
  {{/if}}
  {{#if badges.silver}}
    <g transform="translate({{badgeSilverMarginLeft}}, 0)" class="badge-count">
      <circle fill="var(--color-silver)"></circle>
      <text x="7">{{badges.silver}}</text>
    </g>
  {{/if}}
  {{#if badges.bronze}}
    <g transform="translate({{badgeBronzeMarginLeft}}, 0)" class="badge-count">
      <circle fill="var(--color-bronze)"></circle>
      <text x="7">{{badges.bronze}}</text>
    </g>
  {{/if}}
</g>