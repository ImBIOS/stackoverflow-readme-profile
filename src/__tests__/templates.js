import { renderError, renderProfile, renderProfileSmall } from '../templates'

describe('Templates functions', () => {
  describe('error svg template', () => {
    it('should match the snapshot', () => {
      expect(renderError({error: 'Error: Unable to fetch the user with the id 858453669'})).toMatchSnapshot()
    })
  })

  describe('profile svg template', () => {
    it('should match the snapshot', () => {
      expect(renderProfile({
        avatarLink: 'foobar',
        username: 'foobar',
        reputation: '42k',
        badges: {
          gold: 42,
          silver: 42,
          bronze: 42
        },
        location: 'foobar',
        website: 'foobar'
      })).toMatchSnapshot()
    })
  })

  describe('profile-small svg template', () => {
    it('should match the snapshot', () => {
      expect(renderProfileSmall({
        avatarLink: 'foobar',
        username: 'foobar',
        reputation: '42k',
        badges: {
          gold: 42,
          silver: 42,
          bronze: 42
        },
        location: 'foobar',
        website: 'foobar'
      })).toMatchSnapshot()
    })
  })
})