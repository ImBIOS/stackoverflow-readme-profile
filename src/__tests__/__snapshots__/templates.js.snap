// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates functions error svg template should match the snapshot 1`] = `
"<svg version=\\"1.1\\" baseProfile=\\"full\\" width=\\"300\\" height=\\"86\\" viewBox=\\"0 0 300 86\\" preserveAspectRatio=\\"xMinYMin meet\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"overflow: visible;\\"><style>svg {
      overflow: visible;
    }
    .ctn {
      x: 0;
      y: 0;
      fill: #F8D7DB;
      stroke-width: 2;
      stroke: #AD6E75;
    }
    text {
      fill: #AD6E75;
      font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif;
    }
    tspan {
      font-weight: bold;
    }</style><rect width=\\"300\\" height=\\"86\\" class=\\"ctn\\" rx=\\"5\\" ry=\\"5\\"></rect><text x=\\"11\\" y=\\"11\\" font-weight=\\"bold\\">  <tspan x=\\"11\\" dx=\\"6\\" dy=\\"12\\">Error: Unable to fetch the user with the id 8</tspan>  <tspan x=\\"11\\" dx=\\"6\\" dy=\\"12\\">58453669</tspan>  </text></svg>"
`;

exports[`Templates functions profile svg template should match the snapshot 1`] = `
"<svg version=\\"1.1\\" baseProfile=\\"full\\" width=\\"300\\" height=\\"86\\" viewBox=\\"0 0 300 86\\" preserveAspectRatio=\\"xMinYMin meet\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"overflow: visible;\\"><style> :root { --color-text: #000; --color-bg: #FFFEFE; --color-link: #536675; --color-primary: #000; --color-gold: #FFCC01; --color-silver: #B4B8BC; --color-bronze: #D1A684; --color-so-icon: #E6E5E5; } @keyframes fadeIn { 0% {opacity:0;} 100% {opacity:1;} } svg { overflow: visible; } .animate { animation: fadeIn .5s forwards; opacity: 0; } text { fill: #000; font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif; } .ctn { x: 0; y: 0; fill: var(--color-bg); stroke-width: 2; stroke: var(--color-primary); } .username {
      font-weight: 500;
      font-size: 16px;
    }
    .website, .location {
      font-size: 10px;
    }
    .website {
      fill: var(--color-link);
    }</style><rect width=\\"300\\" height=\\"86\\" class=\\"ctn\\" rx=\\"5\\" ry=\\"5\\"></rect> <style>.so-icon {
    transform: translateX(230px);
  }
  .so-icon path {
    fill: var(--color-so-icon);
  }</style><g class=\\"so-icon\\"><svg aria-hidden=\\"true\\" width=\\"62px\\" viewBox=\\"0 0 25 30\\"><path d=\\"M21 27v-8h3v11H0V19h3v8h18z\\" fill=\\"#BCBBBB\\"></path><path d=\\"M17.1.2 15 1.8l7.9 10.6 2.1-1.6L17.1.2zm3.7 14.7L10.6 6.4l1.7-2 10.2 8.5-1.7 2zM7.2 12.3l12 5.6 1.1-2.4-12-5.6-1.1 2.4zm-1.8 6.8 13.56 1.96.17-2.38-13.26-2.55-.47 2.97zM19 25H5v-3h14v3z\\" fill=\\"#F48024\\"></path></svg></g> <svg x=\\"11\\" y=\\"11\\">  <rect id=\\"imageCtn\\" transform=\\"translate(0,0)\\" width=\\"64\\" height=\\"64\\" rx=\\"5\\"/><clipPath id=\\"clip\\"><use href=\\"#imageCtn\\"/></clipPath><use href=\\"#imageCtn\\"/><image width=\\"64\\" height=\\"64\\" clip-path=\\"url(#clip)\\" href=\\"foobar\\"/>  <g transform=\\"translate(75,13)\\"><text class=\\"username animate\\" transform=\\"translate(0,0)\\">foobar</text><g class=\\"animate\\" transform=\\"translate(0,22)\\" style=\\"font-weight: 800; animation-delay: 0.1s;\\"> <style>.reputation {
    font-weight: 600;
    font-size: 14px;
  }
  .badge-count circle {
    cx: 0;
  }</style><text class=\\"reputation\\">42k</text><g transform=\\"translate(39, 0)\\">  <g class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-gold)\\"></circle><text x=\\"7\\">42</text></g>   <g transform=\\"translate(30, 0)\\" class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-silver)\\"></circle><text x=\\"7\\">42</text></g>   <g transform=\\"translate(60, 0)\\" class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-bronze)\\"></circle><text x=\\"7\\">42</text></g>  </g> </g>  <g class=\\"animate\\" style=\\"animation-delay: 0.2s;\\" transform=\\"translate(0,38)\\"><text class=\\"location\\">foobar</text></g>   <g class=\\"animate\\" style=\\"animation-delay: 0.3s;\\" transform=\\"translate(0,50)\\"><a href=\\"foobar\\" target=\\"_blank\\"><text class=\\"website\\">foobar</text></a></g>  </g></svg></svg>"
`;

exports[`Templates functions profile-small svg template should match the snapshot 1`] = `
"<svg version=\\"1.1\\" baseProfile=\\"full\\" width=\\"240\\" height=\\"57\\" viewBox=\\"0 0 240 55\\" preserveAspectRatio=\\"xMinYMin meet\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"overflow: visible;\\"><style> :root { --color-text: #000; --color-bg: #FFFEFE; --color-link: #536675; --color-primary: #000; --color-gold: #FFCC01; --color-silver: #B4B8BC; --color-bronze: #D1A684; --color-so-icon: #E6E5E5; } @keyframes fadeIn { 0% {opacity:0;} 100% {opacity:1;} } svg { overflow: visible; } .animate { animation: fadeIn .5s forwards; opacity: 0; } text { fill: #000; font: 500 12px 'Segoe UI', Ubuntu, Sans-Serif; } .ctn { x: 0; y: 0; fill: var(--color-bg); stroke-width: 2; stroke: var(--color-primary); } .username {
      font-weight: 500;
      font-size: 16px;
    }</style><rect width=\\"240\\" height=\\"57\\" class=\\"ctn\\" rx=\\"5\\" ry=\\"5\\"></rect> <style>.so-icon {
    transform: translateX(195px);
  }
  .so-icon path {
    fill: var(--color-so-icon);
  }</style><g class=\\"so-icon\\"><svg aria-hidden=\\"true\\" width=\\"40px\\" viewBox=\\"0 0 25 30\\"><path d=\\"M21 27v-8h3v11H0V19h3v8h18z\\" fill=\\"#BCBBBB\\"></path><path d=\\"M17.1.2 15 1.8l7.9 10.6 2.1-1.6L17.1.2zm3.7 14.7L10.6 6.4l1.7-2 10.2 8.5-1.7 2zM7.2 12.3l12 5.6 1.1-2.4-12-5.6-1.1 2.4zm-1.8 6.8 13.56 1.96.17-2.38-13.26-2.55-.47 2.97zM19 25H5v-3h14v3z\\" fill=\\"#F48024\\"></path></svg></g> <svg x=\\"5\\" y=\\"5\\">  <rect id=\\"imageCtn\\" transform=\\"translate(0,0)\\" width=\\"47\\" height=\\"47\\" rx=\\"5\\"/><clipPath id=\\"clip\\"><use href=\\"#imageCtn\\"/></clipPath><use href=\\"#imageCtn\\"/><image width=\\"47\\" height=\\"47\\" clip-path=\\"url(#clip)\\" href=\\"foobar\\"/>  <g transform=\\"translate(55,20)\\"><text class=\\"username animate\\" transform=\\"translate(0,0)\\">foobar</text><g class=\\"animate\\" transform=\\"translate(0,20)\\" style=\\"font-weight: 800; animation-delay: 0.1s;\\"> <style>.reputation {
    font-weight: 600;
    font-size: 14px;
  }
  .badge-count circle {
    cx: 0;
  }</style><text class=\\"reputation\\">42k</text><g transform=\\"translate(39, 0)\\">  <g class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-gold)\\"></circle><text x=\\"7\\">42</text></g>   <g transform=\\"translate(30, 0)\\" class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-silver)\\"></circle><text x=\\"7\\">42</text></g>   <g transform=\\"translate(60, 0)\\" class=\\"badge-count\\"><circle cy=\\"-4\\" r=\\"4\\" fill=\\"var(--color-bronze)\\"></circle><text x=\\"7\\">42</text></g>  </g> </g></g></svg></svg>"
`;
